### ISSUE TITLE
Fix updateAll and deleteAll implementations in all connectors

### ISSUE THREAD

**Issue by bajtos** [strongloop/loopback#1167]
- [x] 1. Add a test to juggler's test suite run by all connectors to check that `updateAll` is implemented and return the number of updated records.
- [x] 2. Fix Memory connector to provide the number of updated records
- [ ] 3. Run the new test in all DB connectors and fix any test failures
  - [x] Verify and fix MongoDB connector changes
  - [x] Verify and fix MySQL connector changes
  - [x] Verify and fix PostgreSQL connector changes
  - [x] Verify and fix MSSQL connector changes
  - [x] Verify and fix Oracle connector changes

**Comment by chandadharap** [comment]
Prioritize these connectors:
Mongo, Mysql, Postgress, MSSQL, Oracle

**Comment by superkhau** [comment]
Connected to strongloop/loopback-datasource-juggler#502

**Comment by superkhau** [comment]
@bajtos For number 3, should I write unit tests for each connector individually? ie) Add two more unit tests directly to`loopback-connector-mongodb`? One that verifies `updateAll` is implemented` and another to verify a count is returned when updating records?

**Comment by bajtos** [comment]
> For number 3, should I write unit tests for each connector individually?

No, add your new tests to [test/manipulation.test.js](https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/manipulation.test.js). This file is run by all connectors as part of their test suite (or at least it should be) via [test/common.batch.js](https://github.com/strongloop/loopback-datasource-juggler/blob/master/test/common.batch.js).

**Comment by bajtos** [comment]
See also https://github.com/strongloop/loopback-datasource-juggler/blob/274a5c778ae0e76403fbb1480c7b144b898c3e18/test/memory.test.js#L404-L427, I am not sure if `dao.js` provides a fallback implementation for connectors that don't implement `updateAll`. If it does, then we should test it too.

**Comment by bajtos** [comment]
It's ok to add the new tests to `test/basic-querying.test.js` as you did in #502, considering that other `updateAll` tests are already living there.

**Comment by superkhau** [comment]
3. Mongo connector test fails with:     

```
Uncaught AssertionError: expected {
[MongoError: '$set' is empty. You must specify a field like so: {$mod: {<field>: ...}}]
  code: 9,
  errmsg: '\'$set\' is empty. You must specify a field like so: {$mod: {<field>: ...}}',
  index: 0,
  name: 'MongoError'
} to not exist
```

**Comment by raymondfeng** [comment]
@superkhau For update, you need to have at least one property in the data arg.

**Comment by superkhau** [comment]
@raymondfeng The test itself is testing for `undefined` though? See https://github.com/strongloop/loopback-datasource-juggler/pull/502/files#diff-404450f6fd92e7f3f2031232213c07c5R981

**Comment by raymondfeng** [comment]
The failing test case should have been fixed in master. You can rebase.

**Comment by superkhau** [comment]
Still failing.

**Comment by bajtos** [comment]
I turns out I need `deleteAll` to report the number of deleted records too. https://github.com/strongloop/loopback-datasource-juggler/pull/511 is implementing the changes in juggler, we need to fix connectors too.

@superkhau would you mind taking care of fixing `deleteAll` in the connectors as a part of your work on `updateAll`, and keeping this GH issue to track both?

/cc @raymondfeng

**Comment by superkhau** [comment]
Sure.

Connect to strongloop/loopback-datasource-juggler#511

**Comment by superkhau** [comment]
Do we expect consistent data types among all connectors? For example:

```
//mongo connector
...update(function(err, count) {
  //count is a number here
}

//mysql connector
...update(function(err, data) {
  //data is an object containing count
}

//data
   { fieldCount: 0,
     affectedRows: 0,
     insertId: 0,
     serverStatus: 34,
     warningCount: 0,
     message: '(Rows matched: 0  Changed: 0  Warnings: 0',
     protocol41: true,
     changedRows: 0 } }
```

Should we be returning custom results based on the type of connector? Or just simply a number (count) of items affected for all connectors to be consistent?

**Comment by bajtos** [comment]
I strongly prefer juggler and connectors to return a consistent value independent on the connector used. In this case just a number (count) of affected rows. The connector-specific code should stay in connectors, juggler should not need to know about different response formats.

We should probably preserve current behaviour for backward compatibility and send the "count" value in a new callback argument, i.e. `cb(null, data, count)`.

@raymondfeng thoughts?

**Comment by bajtos** [comment]
If we decide to go with `cb(null, data, count)`, then I have to rework https://github.com/strongloop/loopback-datasource-juggler/pull/511 to use the same pattern.

**Comment by bajtos** [comment]
> We should probably preserve current behaviour for backward compatibility and send the "count" value in a new callback argument, i.e. cb(null, data, count).

Ok, I see that https://github.com/strongloop/loopback-datasource-juggler/pull/502 is returning an object with a count property to the callback, that looks like a good approach. I'll apply the same principle in https://github.com/strongloop/loopback-datasource-juggler/pull/510

**Comment by raymondfeng** [comment]
I prefer that connectors to call back with single data object. The `count` should be a property of `data` in this case. If existing CRUD apis on `PersistedModel` expects a callback with `count`, it can be handled by dao.js.

**Comment by superkhau** [comment]
:+1:  to single data object. I am modifying all the other connectors to match this approach.

**Comment by superkhau** [comment]
@bajtos @raymondfeng Postgres connector doesn't return number of rows deleted in `executeQuery` for `deleteAll`. It returns:

```
{ command: 'DROP',
  rowCount: NaN,
  oid: null,
  rows: [],
  fields: [],
  _parsers: [],
  RowCtor: null,
  rowAsArray: false,
  _getTypeParser: [Function] }
```

This is kind of a blocker because I'm not sure how to get the count. The ugly option I can think of is to perform a count query (`SELECT COUNT(*) FROM table WHERE "name" = "John"`), then perform the delete query (`DELETE FROM table WHERE "name"="John"`). Can you guys think of any solutions?

**Comment by bajtos** [comment]
What version of PostgreSQL are you using? I am also surprised that the response contains `command: 'DROP'`, I would expect `command: 'DELETE'` instead.

The docs for both [8.2](http://www.postgresql.org/docs/8.2/static/sql-delete.htm) and 9.0 says that

> On successful completion, a DELETE command returns a command tag of the form

> ```
DELETE count
```
> The count is the number of rows deleted. If count is 0, no rows matched the condition (this is not considered an error).

You can try to add a `RETURNING` clause

> If the DELETE command contains a RETURNING clause, the result will be similar to that of a SELECT statement containing the columns and values defined in the RETURNING list, computed over the row(s) deleted by the command.

Related reading: [StackOverflow thread](http://stackoverflow.com/questions/2251567/how-to-get-the-number-of-deleted-rows-in-postgresql) and [PL/pgSQL statements](http://www.postgresql.org/docs/devel/static/plpgsql-statements.html).

**Comment by superkhau** [comment]
https://github.com/strongloop/loopback-connector-postgresql/blob/master/package.json#L18 for the version of the `pg` lib we use.

I'm not sure if just the `pg` lib yet, but I'll try adding the `RETURNING` clause and see how that goes.

**Comment by superkhau** [comment]
@bajtos @raymondfeng FYI, The `RETURNING` clause works. For reference, the clause on SO fails, but the clause at the bottom of [this page](http://www.postgresqltutorial.com/postgresql-delete/) works. tl;dr - `RETURNING *` works, `IS TRUE RETURNING *` doesn't.

**Comment by bajtos** [comment]
Please don't fetch all data from the database via `RETURNING *`, try to limit the amount of data as much as possible, e.g. by returning only a list of IDs via `RETURNING id` or perhaps get only the count via `RETURNING COUNT(*)` if that works.

**Comment by superkhau** [comment]
Makes sense. Limiting with `RETURNING id` instead now as `RETURNING COUNT(*)` doesn't work - aggregate functions not allowed in `RETURNING` clauses.

**Comment by raymondfeng** [comment]
@superkhau I don't think you have to mess around `RETURNING'. Here is what I see from the debug output of Model.deleteAll:
```
loopback:connector:postgresql {"command":"DELETE","rowCount":1,"oid":null,"rows":[],"fields":[],"_parsers":[],"RowCtor":null,"rowAsArray":false} +3ms
null []
```
The `rowCount` is what we need.

**Comment by superkhau** [comment]
@raymondfeng I will double check before messing with the query itself to make sure `rowCount` is there.

**Comment by superkhau** [comment]
@raymondfeng After more digging, the `rowCount` is in the results object of the original query. Unfortunately, we only return [`data.rows`](https://github.com/strongloop/loopback-connector-postgresql/blob/master/lib/postgresql.js#L122). Anyways, I think the root of the issue and my confusion is this:

1. `executeSQL` is low level and called by `prototype.query`, it should return `err` and the `results` object which contains `rowCount` and the `data`
2. `query` should pass that back up to the calling function
3. The calling function (`update/delete`) should massage the data/metadata before returning the results to the caller

tl;dr
We don't return the actual results object in a shared low level function, so the higher level callers don't get to see things like `rowCount` because we only return `data.rows`, etc. I'm proposing we simplify the low level functions to return all data/metadata and only massage in the higher level functions like `deleteAll/updateAll`

**Comment by bajtos** [comment]
> We don't return the actual results object in a shared low level function, so the higher level callers don't get to see metadata like rowCount because we only return data.rows. I'm going to modify executeSQL to return ALL data (which includes metadata like rowCount.

+1

Just be careful to not break the public API exposed by the connector please. If `executeSQL` is not exported then you are fine.

**Comment by raymondfeng** [comment]
@superkhau +1.

**Comment by superkhau** [comment]
@bajtos After speaking with @raymondfeng on hangouts, it was decided that we can just modify the second argument (get rid of `data.rows`) which allows me to use the second arg for all connectors. I am going to modify all the third arg updates to do that instead.

**Comment by bajtos** [comment]
> After speaking with @raymondfeng on hangouts, it was decided that we can just modify the second argument (get rid of data.rows) which allows me to use the second arg for all connectors. I am going to modify all the third arg updates to do that instead.

I am not convinced this is not a BC, but I'll go with whatever decision you have made so that we can get this finished ASAP.

**Comment by superkhau** [comment]
@bajtos Had another meeting with @raymondfeng explaining the BC issue you just mentioned. Gonna update the PostgreSQL connector to return `data.rows` if selecting and `{count: affectedRows}` on `destroyAll/updateAll`. We can discuss any further changes directly there.

**Comment by bajtos** [comment]
@superkhau could you please describe the callback arguments in the documentation explaining how to build a connector? http://docs.strongloop.com/display/LB/Building+a+connector

I am not sure if the API documentation in juggler's `lib/dao.js` was updated to match the changes we made to `update*` and `delete*` callbacks, could you please check that too?

**Comment by bajtos** [comment]
@superkhau for example, updateAll still says that a scalar "count" value is returned: https://github.com/strongloop/loopback-datasource-juggler/blob/e13c0cdc12467aff5f347fa1112c03ba1dcdbb8b/lib/dao.js#L1634 It would be nice to fix the implementation too, it's using a property called `count` instead of `info` now.

What is the final loopback/juggler API that you have arrived at with @raymondfeng? Is the description below correct?

```js
MyModel.updateAll -> cb(null, { count: X });
MyModel.deleteAll -> cb(null, { count: X });
```

**Comment by bajtos** [comment]
> I am not sure if the API documentation in juggler's lib/dao.js was updated to match the changes we made to update* and delete* callbacks, could you please check that too?

I made the changes myself, see  https://github.com/strongloop/loopback-datasource-juggler/pull/540 and https://github.com/strongloop/loopback/pull/1252

**Comment by superkhau** [comment]
>@superkhau could you please describe the callback arguments in the documentation explaining how to build a connector? http://docs.strongloop.com/display/LB/Building+a+connector

Done.
